### PR TITLE
Gamma correction filter for raster layers

### LINKED ISSUE #13512: Implement gamma correction for raster maps
Author Name: **Markus Neteler** (Markus Neteler)
Original Redmine Issue: [3452](https://issues.qgis.org/issues/3452)

Redmine category:rasters
Assignee: nobody -

---
To visually enhance raster maps such as satellite data from files or WMS, support of gamma correction (http://en.wikipedia.org/wiki/Gamma_correction) would be helpful, for example besides the transparency slider.

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Add gamma correction filter pipe for raster layers in addition to the existing brightness, contrast, hue and saturation filters. User can adjust gamma value from the layer properties dialog, using styling panel and/or new buttons on the Raster toolbar.
Peek 2020-06-28 11-04
Default gamma value is 1.0 and it can be adjusted is the range 0.1…10.

Fixes #13512.

Also adds a rendering test for brightness and contrast filter which has no test at all and fixes some deprecation warnings in the Python test for `QgsRasterLayer`.

### CODE REVIEW

**Comment by nyalldawson** [comment]
Nice feature!

I'm curious why this was added as a new filter and not as an extension to the existing brightness/contrast filter? There's perfomance benefits from keeping it in the existing filter if both brightness/contrast AND gamma are changed from non default values -- by splitting them to too different filters we double the allocations and have to iterate over all pixels twice, which is costly.

**Comment by alexbruy** [comment]
@nyalldawson it was added as separate filter as we already have separate hue/saturation and brightness/contrast filters. Also adding gamma to the brightness/contrast filter probably will be a bit confusing without renaming filter class. But I'm fine with merging brightness/contrast and gamma, will work on it ASAP.

**Comment by alexbruy** [comment]
Moved inside brightness/contrast pipe.

File: src/core/raster/qgsbrightnesscontrastfilter.h [hunk 1]
```diff
@@ -49,20 +49,27 @@ class CORE_EXPORT QgsBrightnessContrastFilter : public QgsRasterInterface
     void setContrast( int contrast ) { mContrast = qBound( -100, contrast, 100 ); }
     int contrast() const { return mContrast; }
 
+    void setGamma( double gamma ) { mGamma = qBound( 0.1, gamma, 10.0 ); }
+    double gamma() const { return mGamma; }
```
Comment (nyalldawson) [inline src/core/raster/qgsbrightnesscontrastfilter.h:53]: Don't forget the dox for these!

**Review by nyalldawson — COMMENTED** [review]
Looking good! Only a tiny change needed...

File: src/core/raster/qgsbrightnesscontrastfilter.h (hunk 1 above)
Comment (alexbruy) [reply src/core/raster/qgsbrightnesscontrastfilter.h]: Done

File: python/core/auto_generated/raster/qgsbrightnesscontrastfilter.sip.in [hunk 2]
```diff
@@ -24,27 +24,94 @@ Brightness/contrast filter pipe for rasters.
...
     void setBrightness( int brightness );
+%Docstring
+Set brightness level. Acceptable value range is -255…255
+
+.. seealso:: :py:func:`brightness`
+%End
+
     int brightness() const;
+%Docstring
+Returns currect brightness level.
```
Comment (nyalldawson) [inline python/core/auto_generated/raster/qgsbrightnesscontrastfilter.sip.in:73]: "Current"(elsewhere too!)

**Review by nyalldawson — APPROVED** [review]

File: python/core/auto_generated/raster/qgsbrightnesscontrastfilter.sip.in (hunk 2 above)
Comment (alexbruy) [reply python/core/auto_generated/raster/qgsbrightnesscontrastfilter.sip.in]: Oops. My fault, should be fixed now.

**Comment by neteler** [comment]
Great work, thank you!

**Pull request merged**